Add Sign-up API

prakhar1912 opened the pull request:
The Sign-up API at /api/signup has been added. Here is the functionality:

![signup api](https://user-images.githubusercontent.com/20063271/53132366-850afd80-3595-11e9-8241-3bcbce967f12.png)

prakhar1912 commented:
@nimish-gupta The Travis build fails for this because a `JWT_SECRET` variable is required in the `env`. I have tried adding it in the Travis dashboard, but I don't see it being reflected. This is probably an access right issue. What do you suggest I should do?

nimish-gupta commented:
@prakhar1912 it seems that you haven't added the JWT_SECRET in the dashboard https://www.dropbox.com/s/a51mz9n79au1k87/Screenshot%202019-02-21%2005.22.20.png?dl=0

prakhar1912 commented:
> @prakhar1912 it seems that you haven't added the JWT_SECRET in the dashboard https://www.dropbox.com/s/a51mz9n79au1k87/Screenshot%202019-02-21%2005.22.20.png?dl=0
This is what I see.
![travis](https://user-images.githubusercontent.com/20063271/53133440-2f385480-3599-11e9-8be1-f6e2ff9a3a2e.JPG)

niinpatel commented:
> > @prakhar1912 it seems that you haven't added the JWT_SECRET in the dashboard https://www.dropbox.com/s/a51mz9n79au1k87/Screenshot%202019-02-21%2005.22.20.png?dl=0
> > This is what I see.
> > ![travis](https://user-images.githubusercontent.com/20063271/53133440-2f385480-3599-11e9-8be1-f6e2ff9a3a2e.JPG)

I also see it in my travis dashboard.

nimish-gupta commented:
Still, it is unavailable in the dashboard, can you create a zoom video that how you are reaching your dashboard, and where have you added the secret?

prakhar1912 commented:
> Still, it is unavailable in the dashboard, can you create a zoom video that how you are reaching your dashboard, and where have you added the secret?

Sure.

prakhar1912 commented:
@nimish-gupta Here is the Loom video of the Travis dashboard.
https://www.useloom.com/share/b568ea5c270a43f896750c6d11ebd0ce

nimish-gupta commented:
https://www.useloom.com/share/11cec8f270b045a1896654b3306dc811
Also I re-run the test its good to go :)

prakhar1912 commented:
> https://www.useloom.com/share/11cec8f270b045a1896654b3306dc811
> Also I re-run the test its good to go :)

Thank you so much!

nimish-gupta commented:
@prakhar1912 still I can't figure out that why env were not showing in my dashboard, also can you make a zoom video, that how you reached to your dashboard?

prakhar1912 commented:
> @prakhar1912 still I can't figure out that why env were not showing in my dashboard, also can you make a zoom video, that how you reached to your dashboard?

@niinpatel provided me the link.

niinpatel commented:
> > @prakhar1912 still I can't figure out that why env were not showing in my dashboard, also can you make a zoom video, that how you reached to your dashboard?
> 
> @niinpatel provided me the link.

I just went to https://travis-ci.org/pesto-students/rattle-battle, then settings and environment variables. 

Maybe its a permissions issue?

niinpatel commented:
Okay, I figured out. We had to go to travis-ci**.com** . The **.org** version is deprecated. 
🤦

nimish-gupta commented on server/__tests__/index.test.js:
```diff
@@ -1,18 +1,17 @@
-import server from '../src/index';
+import app from '../src/app/app';
 
+const { TEST_PORT = 8888 } = process.env;
```
Please import env variables from a common place, check [this](https://stackoverflow.com/a/48828102/10322565) for further info

prakhar1912 replied:
Okay, I will make this change.

nimish-gupta commented on server/__tests__/signupAPI.test.js:
```diff
@@ -0,0 +1,98 @@
+import app from '../src/app/app';
+import getMockUser from '../src/utils/testUtils';
+
+const { TEST_PORT = 8888, DB_URL = 'mongodb://localhost:27017/testdb' } = process.env;
```
Please import env variables from a common place, check [this](https://stackoverflow.com/a/48828102/10322565) for further info

nimish-gupta commented on server/src/controllers/userController.js:
```diff
@@ -0,0 +1,29 @@
+  });
+  return newUser
+    .save()
+    .then(() => {
```
Why not async/await?

prakhar1912 replied:
I used .then and .catch just to have a Promise based workflow. I'd like to keep it like that, unless you have a reason for going with async/await?

nimish-gupta replied:
Check [this](https://hackernoon.com/6-reasons-why-javascripts-async-await-blows-promises-away-tutorial-c7ec10518dd9) for async await.
Also its a new feature in ES so you must upgrade your codebase to latest specs :)

prakhar1912 replied:
Okay, I will update this. Thank you for the suggestion.

nimish-gupta commented on server/src/database/models/user.js:
```diff
@@ -0,0 +1,74 @@
+  const token = jwt.sign(
+    {
+      /* eslint-disable-next-line no-underscore-dangle */
+      _id: user._id,
```
Mongoose already adds id getter in the object, check [this](https://stackoverflow.com/a/34424834/10322565) for further info.
Also why `_id`?

prakhar1912 replied:
Okay, I will fix this.

nimish-gupta commented on server/src/database/models/user.js:
```diff
@@ -0,0 +1,74 @@
+ * Model pre-save hook to hash the password if it is changed
+ */
+userSchema.pre('save', function preSaveCheck(next) {
+  const user = this;
```
So as a personal point of view, try to avoid hooks. I am not saying that hooks are a bad pattern, but it leads to hidden errors like it basically obstructs the natural flow of program i.e you write a command `User.save` and you will be thinking that only it will work, but in really the first hook will be called and then save will be performed.
So better option you just create a function that will hash the password for you and then you can use it

prakhar1912 replied:
I did that because at some point we will give the user the option to change their password. So, this just makes for cleaner flow.

nimish-gupta replied:
From that function you can again call the hash pwd function. And you will know that you have called that function instead of hooks which will happen on their own.

prakhar1912 replied:
Okay, got it. I will implement it without the hook.

nimish-gupta commented on server/src/routes/routes.js:
```diff
@@ -1,10 +1,14 @@
 router.get('/ping', (req, res) => {
   res.json({ status: 'working' });
 });
+router.post('/api/signup', userValidator, userFieldsErrors, userController.createNewUser);
```
Its good call that you are using userValidator and userFieldErrors middleware over here.
But if you are adding restrictions on the db fields then it's better that db checks the object itself and then throws the error.
You can provide all the validations like length, required with the mongoose schema itself, you don't have to check it upfront with another library

niinpatel replied:
I actually like this approach better, because it enables us to send user-friendly error messages to client.

I mean, if we let mongodb throw an error, we'll have to parse the error message and send the user-friendly message to client. Or send something generic like "there was an error".

Also, if there are more than one validation errors, user can see all of them at once because we're sending errors array. This is not possible in database level validation I think.

prakhar1912 replied:
I agree with Nitin.

nimish-gupta commented:
> Okay, I figured out. We had to go to travis-ci **.com** . The **.org** version is deprecated. 🤦

.org is for private repos @niinpatel

prakhar1912 commented:
@nimish-gupta Thank you for the review. I will make the changes you've requested, and I hope I answered all the questions you had.

prakhar1912 commented:
> Savage PR :)
> It's good how you use the test server for testing. Some questions though then it LGTM.
> 
> * Why there is an extra blank line at top of each file?
> * Since for every server test you might be starting the server and connecting server and then, in the end, you might be closing the connection, check [this](https://jestjs.io/docs/en/setup-teardown) that how jest provides this functionality.
> * Please import env variables from a common place, check [this](https://stackoverflow.com/a/48828102/10322565) for further info.

I will remove the blank line at the top of each file.

niinpatel commented on server/src/database/connection.js:
```diff
@@ -0,0 +1,8 @@
+
+const { DB_URL } = process.env;
+
+mongoose.connect(DB_URL, { useNewUrlParser: true, useCreateIndex: true });
```
The connect() function also accepts a callback parameter in third argument. see: https://mongoosejs.com/docs/connections.html#callback

 You can use it to log if there are any errors connecting to database. something like: 

`(err) => if(err) console.log(err) else console.log('connected to mongodb database');`

vkbinfo replied:
Also, do the logging in a file for error issues.

niinpatel commented on server/src/validators/userValidator.js:
```diff
@@ -0,0 +1,17 @@
+    .isEmail()
+    .withMessage('must be a valid email'),
+  check('username')
+    .optional()
```
username is required in schema, then why optional here??

prakhar1912 replied:
This is because I will use the same validator for the login API, and username will be optional there. I should probably make this a function that take a parameter denoting if username is optional or not. Thoughts?

niinpatel replied:
Yeah, we only want username to be option when logging in and not when signing up. So, that would be a good idea. 

Or we catch error thrown by mongodb and send response to client accordingly. But first approach seems better to me.

prakhar1912 replied:
Okay, I will do that in the login PR.

prakhar1912 commented:
@nimish-gupta I have made the following changes as requested:
1) Removed the empty line at the beginning of files.
2) Used async/await instead of .then
3) Accessing env variables from a file called `appConstants`
4) Removed the pre save hook for hashing passwords
5) Using the virtual getter for user ID

nimish-gupta commented:
> * Since for every server test you might be starting the server and connecting server and then, in the end, you might be closing the connection, check [this](https://jestjs.io/docs/en/setup-teardown) that how jest provides this functionality.

@prakhar1912 have you checked this?

prakhar1912 commented:
> > * Since for every server test you might be starting the server and connecting server and then, in the end, you might be closing the connection, check [this](https://jestjs.io/docs/en/setup-teardown) that how jest provides this functionality.
> 
> @prakhar1912 have you checked this?

Yes, I am already using `beforeEach` and `afterEach`. Do you have a different flow in mind?

nimish-gupta commented:
@prakhar1912  have a look at these sites [1](https://jestjs.io/docs/en/configuration.html#globalsetup-string), [2](https://jestjs.io/docs/en/configuration.html#globalteardown-string) and [3](https://jestjs.io/docs/en/configuration.html#setupfiles-array)

prakhar1912 commented:
> @prakhar1912 have a look at these sites [1](https://jestjs.io/docs/en/configuration.html#globalsetup-string), [2](https://jestjs.io/docs/en/configuration.html#globalteardown-string) and [3](https://jestjs.io/docs/en/configuration.html#setupfiles-array)

@nimish-gupta Okay. Thank you for those resources. Now I think I know what to do.

prakhar1912 commented:
> @prakhar1912 have a look at these sites [1](https://jestjs.io/docs/en/configuration.html#globalsetup-string), [2](https://jestjs.io/docs/en/configuration.html#globalteardown-string) and [3](https://jestjs.io/docs/en/configuration.html#setupfiles-array)

@nimish-gupta Here is what @niinpatel and I found about the Jest config:

Globals defined in `globalSetup` cannot be accessed in the the test files themselves. So, this is not suited for the API test, as we need to running `testServer` to work with `supertest`. Please advise.

nimish-gupta commented:
You have to add the server in the global variable and have to use that
variable to start or close the server.

On Thu, Feb 21, 2019 at 11:22 PM Prakhar Yadav <notifications@github.com>
wrote:

> @prakhar1912 <https://github.com/prakhar1912> have a look at these sites 1
> <https://jestjs.io/docs/en/configuration.html#globalsetup-string>, 2
> <https://jestjs.io/docs/en/configuration.html#globalteardown-string> and 3
> <https://jestjs.io/docs/en/configuration.html#setupfiles-array>
>
> @nimish-gupta <https://github.com/nimish-gupta> Here is what @niinpatel
> <https://github.com/niinpatel> and I found about the Jest config:
>
> Globals defined in globalSetup cannot be accessed in the the test files
> themselves. So, this is not suited for the API test, as we need to running
> testServer to work with supertest. Please advise.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pesto-students/rattle-battle/pull/41#issuecomment-466098522>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMWZ091cPvQGzOCihQaIN-WAM3hs3zLiks5vPtzIgaJpZM4bGVS0>
> .
>

vkbinfo commented on server/__tests__/signupAPI.test.js:
```diff
@@ -0,0 +1,96 @@
+  mockUser = getMockUser();
+  testServer = app.listen(TEST_PORT, () => {
+    request = supertest(testServer);
+    return done();
```
Should we use return here? because it is an anonymous function. we are not assigning the result to anything. @nimish-gupta could you give your thoughts here?

nimish-gupta replied:
you can just call done() or return done(), it doesn't make much sense over here. We just need to call done which is imp.

prakhar1912 replied:
It seems better to me if functions have a return statement in them. I can change it to just `done()`. Do you want me to?

prakhar1912 replied:
I'll just remove the return there.

vkbinfo commented on server/__tests__/signupAPI.test.js:
```diff
@@ -0,0 +1,96 @@
+      .expect(200)
+      .end((err, res) => {
+        const createdUser = res.body;
+        expect(createdUser).to.have.keys('id', 'username');
```
Shouldn't we check the username value here with the mock user?

prakhar1912 replied:
I have added this. Thanks @vkbinfo

vkbinfo commented on server/__tests__/signupAPI.test.js:
```diff
@@ -0,0 +1,96 @@
+      .expect(422)
+      .end(() => done());
+  });
+  it('gives error messages when email or username or password are missing', (done) => {
```
I think you can check status code and error message in one test.

prakhar1912 replied:
@niinpatel @nimish-gupta Thoughts?

nimish-gupta commented:
@prakhar1912 check [this](https://jestjs.io/docs/en/mongodb) for global setup and teardown

prakhar1912 commented:
> @prakhar1912 check [this](https://jestjs.io/docs/en/mongodb) for global setup and teardown

@nimish-gupta  That is the problem. If I create a server in `globalSetup`, I cannot access that global variable in the test file. I can only use that variable to close the server in the `globalTeardown`. I tried assigning a running server to `global.TESTSERVER`, but the test file could not access it.

nimish-gupta commented:
@prakhar1912 You have to extend the jest environment using [this](https://jestjs.io/docs/en/configuration.html#testenvironment-string) you have to right the config properties in a file under setup then from here you have to read properties and create a global flag under jest which will be available in all tests. 

I know it is a bit complicated, feel free to merge this PR , we can discuss this thing tomorrow in person as I will be in office. and create a separate PR for that.

prakhar1912 commented:
> @prakhar1912 You have to extend the jest environment using [this](https://jestjs.io/docs/en/configuration.html#testenvironment-string) you have to right the config properties in a file under setup then from here you have to read properties and create a global flag under jest which will be available in all tests.
> 
> I know it is a bit complicated, feel free to merge this PR , we can discuss this thing tomorrow in person as I will be in office. and create a separate PR for that.

Okay. I'd like to complete Jest configuration before merging this PR. I know I can figure it out.

prakhar1912 commented:
@nimish-gupta The testing change has been made. Awaiting review before merging.
